### ISSUE TITLE
Sensitive data masking for debug logging

### ISSUE THREAD

**Issue by jamesratzlaff** [dependency-check/DependencyCheck#2147]
#### (I can add an actual git issue, if you would like me to)
**TLDR;** The methods Settings.logProperties and Settings.setString can expose sensitive data.

The debug output is very helpful when troubleshooting in my remote dev environment, *but* there are certain things I never want to be logged, such as tokens to various APIs. When calling Settings.logProperties it does a good job of masking any property whose key contains "password" when listing the values, but it misses other items that are equivalent to passwords such as tokens.

In my case, I'm currently using property placeholders so I'm not storing/revealing sensitive data in plain-text anywhere. In code, I'm post processing those values and programatically setting the resolved/actual value on the Settings object. This whole "going-the-extra-mile" is made moot  whenever setString is called on the Settings object. It will log the key and value of whatever I'm setting, **even my sensitive values**. 

### Proposed solution
Instead having hard-coded value of which keys to mask-out values of, would it be possible to add a settings property that allows something along the lines of a glob,regex,or straight key name? These values could be used to check if a given setting is considered to be sensitive and 
 masked in the log output while other information that is less sensitive can still be seen.

Example (using a glob-style pattern which can be converted to a regex in a similar fashion as sun.nio.fs.Globs)

in dependencycheck.properties
```
settings.mask=*password*,*.token
```
then in Settings.java
```java

/**
 * could also be a single Predicate<String>by 
 * Predicate.or-ing each expression 
 */
private List<Predicate<String>> maskedKeys; 


// ..read in settings.mask and convert to some 
// a predicate or predicate-like object(s)
 //...

/**
 *Check if a given key is considered to have a value with sensitive data
 */
private boolean isKeyMasked(@NotNull String key){
    return maskedKeys.stream().anyMatch(maskExp->maskExp.test(key));
}

//...


private String getPrintableValue(@NotNull String key, String value){
    String printableValue = null;
    if(value!=null){
        printableValue=isKeyMasked(key)?"********":value;
    }
    return printableValue;
}
/**
 * Logs the properties. This will not log any properties that contain
 * 'password' in the key.
 *
 * @param header the header to print with the log message
 * @param properties the properties to log
 */
private void logProperties(@NotNull final String header, @NotNull final Properties properties) {
        if (LOGGER.isDebugEnabled()) {
            final StringWriter sw = new StringWriter();
            try (PrintWriter pw = new PrintWriter(sw)) {
                pw.format("%s:%n%n", header);
                final Enumeration<?> e = properties.propertyNames();
                while (e.hasMoreElements()) {
                    final String key = (String) e.nextElement();
                    final String value = getPrintableValue(key,properties.getProperty(key));
                    if (value != null) {
                        pw.format("%s='%s'%n", key, value);
                    }
                }
            }
            pw.flush();
            LOGGER.debug(sw.toString());
        }
    }
}
//...
public void setString(@NotNull final String key, @NotNull final String value) {
    props.setProperty(key, value);
    LOGGER.debug("Setting: {}='{}'", key, getPrintableValue(key,value));
}
```
This will allow debug out with out revealing sensitive data.

**Comment by jeremylong** [comment]
Good point - we will make sure to put this in the next release.
